mypy/suggestions.py

```diff
@@ -778,6 +787,25 @@ def count_errors(msgs: List[str]) -> int:
 T = TypeVar('T')
 
 
+def merge_callables(t: CallableType, s: CallableType) -> CallableType:
+    """Merge two callable types in a way that prefers dropping Anys.
+
+    This is implemented by doing a meet on both the arguments and the return type,
+    since meet(t, Any) == t.
+
+    This won't do perfectly with complex compound types (like
+    callables nested inside), but it does pretty well.
+    """
+
+    # We don't want to ever squash away optionals while doing this, so set
+    # strict optional to be true always
+    with strict_optional_set(True):
+        arg_types = []  # type: List[Type]
+        for i in range(len(t.arg_types)):
+            arg_types.append(meet_types(t.arg_types[i], s.arg_types[i]))
```

Comment:
I think this may be a bad idea, technically you should use join here (except `join(C, Any)` is `Any`, while I used to think it should be `C`, see https://github.com/python/mypy/issues/3194).

I could imagine many situations where attempting to improve type for an untyped argument this will cause bad types for existing precise types (like dropping items from a union etc).

A potential solution is to add a flag to `join_types` like `precise=True`. This should be pretty easy, but may cause a bit of churn (every call to `join_types()` in `TypeJoinVisitor` should be replaced with `self.join_types()`).

(Also a related note we should fix `mypy.subtypes.is_more_precise()` at some point.)

Reply:
Oh that's a good catch. I'm going to need to rework the merging a bunch.